File: leetcode/java/1930_unique_length_3_palindromic_subsequences.java
/**
 * 1930. Unique Length-3 Palindromic Subsequences
 * Problem Link: https://leetcode.com/problems/unique-length-3-palindromic-subsequences/
 * Solution Link (Personal):
 */


class Solution {
    public int countPalindromicSubsequence(String s) {
        int[] firstOccurrence = new int[26];
        int[] lastOccurrence = new int[26];
        Arrays.fill(firstOccurrence, Integer.MAX_VALUE);

        for (int i = 0; i < s.length(); i++) {
            firstOccurrence[s.charAt(i) - 'a'] = Math.min(firstOccurrence[s.charAt(i) - 'a'], i);
            lastOccurrence[s.charAt(i) - 'a'] = i;
        }

        int count = 0;
        for (int i = 0; i < 26; i++) {
            if (firstOccurrence[i] < lastOccurrence[i]) {
                Set<Character> set = new HashSet<>();
                for (int j = firstOccurrence[i] + 1; j < lastOccurrence[i]; j++) {
                    set.add(s.charAt(j));
                }
                count += set.size();
            }
        }

        return count;
    }
}

File: leetcode/python3/1930_unique_length_3_palindromic_subsequences.py
"""
 * 1930. Unique Length-3 Palindromic Subsequences 
 * Problem Link: https://leetcode.com/problems/unique-length-3-palindromic-subsequences/
 * Solution Link (Personal):
"""

class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        firstOccurrence = {ch: float('inf') for ch in 'abcdefghijklmnopqrstuvwxyz'}
        lastOccurrence = {ch: -1 for ch in 'abcdefghijklmnopqrstuvwxyz'}

        for i, ch in enumerate(s):
            firstOccurrence[ch] = min(firstOccurrence[ch], i)
            lastOccurrence[ch] = i

        count = 0
        for ch in 'abcdefghijklmnopqrstuvwxyz':
            if firstOccurrence[ch] < lastOccurrence[ch]:
                count += len(set(s[firstOccurrence[ch]+1:lastOccurrence[ch]]))

        return count